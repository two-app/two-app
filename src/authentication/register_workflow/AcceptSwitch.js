import {StyleSheet, Switch, Text, View} from "react-native";
import PropTypes from "prop-types";
import Colors from "../../Colors";
import React, {useState} from "react";

const AcceptBox = ({children, onEmit}) => {
    const [accepted, setAccepted] = useState(false);
    return (
        <View style={styles.container}>
            <Text style={[styles.condition, accepted ? styles.acceptedCondition : undefined]}>{children}</Text>
            <View style={styles.switchContainer}>
                <Switch style={styles.switch}
                        value={accepted}
                        onValueChange={setAccepted}
                />
            </View>
        </View>
    );
};

AcceptBox.propTypes = {
    children: PropTypes.any,
    onEmit: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
    container: {
        flexDirection: "row",
        flex: 1,
        justifyContent: "space-between",
        alignItems: "center",
        backgroundColor: Colors.LIGHT,
        borderRadius: 10,
        padding: 20,
        marginTop: 20
    },
    condition: {
        color: Colors.DARK,
        width: "80%"
    },
    acceptedCondition: {
        color: Colors.FADED
    },
    switchContainer: {
        width: "20%",
        flex: 1,
        alignItems: "center",
        justifyContent: "center"
    },
    switch: {
        transform: [{scaleX: .8}, {scaleY: .8}]
    }
});

export default AcceptBox;