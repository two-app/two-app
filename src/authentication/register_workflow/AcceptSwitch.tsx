import {StyleSheet, Switch, Text, View} from 'react-native';
import PropTypes from 'prop-types';
import React, {useState} from 'react';

import Colors from '../../Colors';

type AcceptBoxProps = {
  children?: any;
  onEmit: any;
  required?: boolean;
  accessibilityHint: string;
};

const AcceptBox = ({
  children,
  onEmit,
  required,
  accessibilityHint,
}: AcceptBoxProps) => {
  const [accepted, setAccepted] = useState(false);

  return (
    <View
      style={[
        styles.container,
        required ? styles.required : undefined,
        accepted ? styles.accepted : undefined,
      ]}
      data-testid="container">
      <Text
        style={[styles.condition, accepted ? styles.acceptedText : undefined]}>
        {children}
      </Text>
      <View style={styles.switchContainer}>
        <Switch
          style={styles.switch}
          value={accepted}
          onValueChange={v => {
            setAccepted(v);
            onEmit(v);
          }}
          accessibilityHint={accessibilityHint}
        />
      </View>
    </View>
  );
};

AcceptBox.propTypes = {
  onEmit: PropTypes.func.isRequired,
  isRequired: PropTypes.bool,
};

AcceptBox.defaultProps = {
  isRequired: false,
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    borderRadius: 10,
    borderColor: Colors.LIGHT,
    borderWidth: 1,
    padding: 20,
    marginTop: 20,
  },
  required: {
    borderColor: Colors.DARK,
  },
  accepted: {
    borderColor: Colors.VALID_GREEN_DARK,
    backgroundColor: Colors.VALID_GREEN,
  },
  acceptedText: {
    color: 'white',
  },
  condition: {
    color: Colors.DARK,
    width: '80%',
  },
  switchContainer: {
    width: '20%',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  switch: {
    transform: [{scaleX: 0.8}, {scaleY: 0.8}],
  },
});

export default AcceptBox;
