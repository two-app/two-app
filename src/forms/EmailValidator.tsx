/**
 * @param email to verify.
 * @return {boolean} true if the email is valid. False if not, or empty.
 */
function validateEmail(email: string) {
  const re = /^\S+@\S+[.][0-9a-z]+$/;
  return re.test(email);
}

export default {validateEmail};
