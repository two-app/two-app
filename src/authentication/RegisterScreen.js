// @flow

import React from "react";
import {ScrollView} from "react-native";
import Input from "../forms/Input";
import {Wrapper} from "../views/View";
import EmailValidator from "../forms/EmailValidator";
import LogoHeader from "./LogoHeader";
import SubmitButton from "../forms/SubmitButton";

const RegisterScreen = ({navigation}) => <Wrapper>
    <ScrollView style={{paddingLeft: '10%', paddingRight: '10%', paddingBottom: '10%'}}>
        <LogoHeader heading="Sign Up"/>

        <Input attributes={{placeholder: "First Name", autoCompleteType: "name"}}
               isValid={v => v.length > 2} label={"First Name"}/>

        <Input attributes={{placeholder: "Last Name"}}
               isValid={v => v.length > 2} label={"Last Name"}/>

        <Input attributes={{placeholder: "you@email.com", autoCompleteType: "email"}}
               isValid={v => EmailValidator.validateEmail(v)} label={"Email"}/>

        <Input attributes={{placeholder: "Secure Password", autoCompleteType: "password", secureTextEntry: true}}
               isValid={v => v.length > 3}
               label={"Password"}/>

        <SubmitButton text="Join Two" onSubmit={() => navigation.navigate("AcceptTermsScreen")} id="submit"/>
    </ScrollView>
</Wrapper>;

RegisterScreen.navigationOptions = {
    title: 'Sign In',
    header: null
};

export default RegisterScreen;