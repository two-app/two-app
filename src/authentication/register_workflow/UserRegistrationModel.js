// @flow

import EmailValidator from "../../forms/EmailValidator";

/**
 * Blank UserRegistration object with Flow types.
 * @type {{firstName: null, lastName: null, password: null, email: null}}
 */
const UserRegistration: {
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    acceptedTerms: boolean,
    ofAge: boolean,
    receivesEmails: boolean
} = {
    firstName: null,
    lastName: null,
    email: null,
    password: null,
    acceptedTerms: false,
    ofAge: false,
    receivesEmails: false
};

const isFirstNameValid = firstName => firstName != null && firstName.length > 1;
const isLastNameValid = lastName => lastName != null && lastName.length > 1;
const isEmailValid = email => email != null && EmailValidator.validateEmail(email);
const isPasswordValid = password => password != null && password.length > 3;

/**
 * @param ur {UserRegistration} to validate
 * @returns {boolean} true if the user supplied information is valid. Does not check terms & age.
 */
const isUserRegistrationValid = (ur: UserRegistration) => (
    isFirstNameValid(ur.firstName) &&
    isLastNameValid(ur.lastName) &&
    isEmailValid(ur.email) &&
    isPasswordValid(ur.password)
);

export default {isFirstNameValid, isLastNameValid, isEmailValid, isPasswordValid, isUserRegistrationValid};
export {UserRegistration};