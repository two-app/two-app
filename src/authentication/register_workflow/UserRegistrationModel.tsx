import "react-native-get-random-values";
import uuid from "uuidv4";

import EmailValidator from "../../forms/EmailValidator";

class UserRegistration {
  uid = uuid();
  firstName = "";
  lastName = "";
  email = "";
  password = "";
  acceptedTerms = false;
  ofAge = false;
  receivesEmails = false;
}

const isFirstNameValid = (firstName: string) =>
  firstName !== "" && firstName.length > 1;
const isLastNameValid = (lastName: string) =>
  lastName !== "" && lastName.length > 1;
const isEmailValid = (email: string) =>
  email !== "" && EmailValidator.validateEmail(email);
const isPasswordValid = (password: string) =>
  password !== "" && password.length >= 6;

const isUserRegistrationValid = (ur: UserRegistration): boolean =>
  isFirstNameValid(ur.firstName) &&
  isLastNameValid(ur.lastName) &&
  isEmailValid(ur.email) &&
  isPasswordValid(ur.password);

export default {
  isFirstNameValid,
  isLastNameValid,
  isEmailValid,
  isPasswordValid,
  isUserRegistrationValid,
};
export { UserRegistration };
