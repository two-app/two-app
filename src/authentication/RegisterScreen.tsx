import React, {useState} from 'react';
import Input from '../forms/Input';
import {WrapperContainer} from '../views/View';
import LogoHeader from './LogoHeader';
import SubmitButton from '../forms/SubmitButton';
import UserRegistrationModel, {UserRegistration} from './register_workflow/UserRegistrationModel';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootStackParamList} from '../../Router';
import { Text, View } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import Colors from '../Colors';
import { CommonActions } from '@react-navigation/native';

type RegisterScreenProps = {
    navigation: StackNavigationProp<RootStackParamList, 'RegisterScreen'>
};

const RegisterScreen = ({navigation}: RegisterScreenProps) => {
    const [userRegistration, setUserRegistration] = useState(new UserRegistration());

    const navigateToLogin = () => {
        navigation.dispatch(
            CommonActions.reset({
                index: 0,
                routes: [{ name: 'LoginScreen' }]
            })
        );
    }

    return (
        <WrapperContainer>
            <LogoHeader heading="Create Account"/>

            <Input attributes={{placeholder: 'First Name', autoCompleteType: 'name'}}
                   isValid={UserRegistrationModel.isFirstNameValid} label={'First Name'}
                   onChange={firstName => setUserRegistration({...userRegistration, firstName})}
            />

            <Input attributes={{placeholder: 'Last Name'}}
                   isValid={UserRegistrationModel.isLastNameValid} label={'Last Name'}
                   onChange={lastName => setUserRegistration({...userRegistration, lastName})}
            />

            <Input attributes={{placeholder: 'you@email.com', autoCompleteType: 'email', autoCapitalize: 'none'}}
                   isValid={UserRegistrationModel.isEmailValid} label={'Email'}
                   onChange={email => setUserRegistration({...userRegistration, email})}
            />

            <Input attributes={{placeholder: 'Secure Password', autoCompleteType: 'password', secureTextEntry: true}}
                   isValid={UserRegistrationModel.isPasswordValid}
                   label={'Password'}
                   onChange={password => setUserRegistration({...userRegistration, password})}
            />

            <SubmitButton text="Join Two"
                          onSubmit={() => navigation.navigate('AcceptTermsScreen', {userRegistration})}
                          disabled={!UserRegistrationModel.isUserRegistrationValid(userRegistration)}
            />

            <View style={{flexDirection: 'row', marginTop: 10}}>
                <Text style={{color: Colors.REGULAR}}>Already a member?</Text>
                <TouchableOpacity style={{marginLeft: 5}} onPress={navigateToLogin}>
                    <Text style={{fontWeight: "bold", color: Colors.DARK}}>Sign In</Text>
                </TouchableOpacity>
            </View>
        </WrapperContainer>
    );
};

RegisterScreen.navigationOptions = {
    title: 'Sign In',
    header: null
};

export default RegisterScreen;