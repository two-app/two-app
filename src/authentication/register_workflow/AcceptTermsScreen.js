// @flow

import React, {useState} from "react";
import {ActivityIndicator, Dimensions, StyleSheet, Text, View} from "react-native";
import AcceptBox from "./AcceptSwitch";
import LogoHeader from "../LogoHeader";
import SubmitButton from "../../forms/SubmitButton";
import {WrapperContainer} from "../../views/View";
import Colors from "../../Colors";
import {UserRegistration} from "./UserRegistrationModel";
import {NavigationActions, StackActions} from "react-navigation";
import {connect} from "react-redux";
import {storeUser} from "../UserReducer";
import AuthenticationService from "../AuthenticationService";

const AcceptTermsScreen = ({navigation, storeUser}) => {
    const [userRegistration: UserRegistration, setUserRegistration] = useState(navigation.getParam("userRegistration"));
    const validAgreedState = userRegistration.acceptedTerms && userRegistration.ofAge;

    const [submitted, setSubmitted] = useState(false);
    const [registrationError, setRegistrationError] = useState(null);

    const navigateToConnectCodeScreen = () => navigation.dispatch(
        StackActions.reset({
            index: 0,
            actions: [NavigationActions.navigate({routeName: 'ConnectCodeScreen'})]
        })
    );

    if (submitted) {
        AuthenticationService.registerUser(userRegistration).then(response => {
            storeUser(response.user);
            navigateToConnectCodeScreen();
        }).catch((e: Error) => {
            setSubmitted(false);
            setRegistrationError(e.message);
        });
    }

    return (<>
        {submitted && <View style={styles.overlay}>
            <ActivityIndicator size="small" color="black" style={styles.overlayIndicator}/>
        </View>}
        <WrapperContainer>
            <LogoHeader heading="Terms & Conditions"/>
            <AcceptBox onEmit={acceptedTerms => setUserRegistration({...userRegistration, acceptedTerms})}
                       id="terms"
                       required>
                I agree to the <Text style={{fontWeight: "bold"}}>Terms & Conditions</Text> and
                <Text style={{fontWeight: "bold"}}> Privacy Policy.</Text>
            </AcceptBox>
            <AcceptBox onEmit={ofAge => setUserRegistration({...userRegistration, ofAge})}
                       id="age"
                       required>
                I am over the age of 16.
            </AcceptBox>
            <AcceptBox onEmit={receivesEmails => setUserRegistration({...userRegistration, receivesEmails})}>
                I agree to occasionally receive emails from Two.
            </AcceptBox>

            <SubmitButton onSubmit={() => setSubmitted(true)} text="Accept" disabled={!validAgreedState}/>
            {registrationError && <Text style={styles.error} id="error-message">{registrationError}</Text>}
        </WrapperContainer>
    </>);
};

AcceptTermsScreen.navigationOptions = {
    title: 'Terms & Conditions',
    header: null
};

const styles = StyleSheet.create({
    overlay: {
        flex: 1,
        position: "absolute",
        opacity: 0.5,
        backgroundColor: 'white',
        width: Dimensions.get("window").width,
        height: Dimensions.get("window").height,
        zIndex: 2,
        alignItems: "center",
        justifyContent: "center"
    },
    overlayIndicator: {
        zIndex: 3
    },
    error: {
        color: Colors.DARK_SALMON
    }
});

export default connect(null, {storeUser})(AcceptTermsScreen);
export {AcceptTermsScreen};