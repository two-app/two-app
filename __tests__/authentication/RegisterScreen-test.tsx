import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {shallow, ShallowWrapper} from 'enzyme';
import RegisterScreen from '../../src/authentication/RegisterScreen';
import EmailValidator from '../../src/forms/EmailValidator';
import {UserRegistration} from '../../src/authentication/register_workflow/UserRegistrationModel';

test('should maintain snapshot', () => expect(renderer.create(<RegisterScreen
    navigation={{} as any}/>)
).toMatchSnapshot());

describe('Navigation Tests', () => {
    test('it should navigate when Join Two is clicked', () => {
        const navigate = jest.fn();

        shallow(<RegisterScreen navigation={{navigate} as any}/>).find('SubmitButton')
            .prop<() => void>('onSubmit')();

        expect(navigate).toHaveBeenCalledWith('AcceptTermsScreen', {userRegistration: new UserRegistration()});
    });
});

describe('Form Validation Tests', () => {

    let wrapper: ShallowWrapper;

    beforeEach(() => wrapper = shallow(<RegisterScreen navigation={{} as any}/>));

    describe('First Name', () => {
        test('invalid with one character', () => expectInput('First Name', 'a').toBe(false));
        test('valid with three characters', () => expectInput('First Name', 'abc').toBe(true));
    });

    describe('Last Name', () => {
        test('invalid with 1 character', () => expectInput('Last Name', 'a').toBe(false));
        test('valid with 3 characters', () => expectInput('Last Name', 'abc').toBe(true));
    });

    describe('Email', () => {
        test('delegates to validEmail fn', () => {
            EmailValidator.validateEmail = jest.fn().mockReturnValue(true);
            expectInput('Email', 'admin@two.com');
            expect(EmailValidator.validateEmail).toHaveBeenCalledTimes(1);
            expect(EmailValidator.validateEmail).toHaveBeenCalledWith('admin@two.com');
        });
    });

    describe('Password', () => {
        test('invalid with < 3 characters', () => expectInput('Password', 'abc').toBe(false));
        test('valid with > 3 characters', () => expectInput('Password', 'abcd').toBe(true));
        test('valid with secure password', () => expectInput('Password', 'P?4Ot2ONz:IJO&%U').toBe(true));
    });

    describe('Fully Valid Form', () => {
        test('should enable the submit button', () => {
            expect(wrapper.find('SubmitButton').prop('disabled')).toBe(true);

            changeInput('First Name', 'Gerry');
            changeInput('Last Name', 'Fletcher');
            EmailValidator.validateEmail = jest.fn().mockReturnValue(true);
            changeInput('Email', 'admin@two.com');
            changeInput('Password', 'P?4Ot2ONz:IJO&%U');

            expect(wrapper.find('SubmitButton').prop('disabled')).toBe(false);
        });
    });

    const expectInput = (label: string, value: string) => {
        const input = wrapper.find(`Input[label='${label}']`).first();
        const isValidFunction = input.prop<(v: string) => boolean>('isValid');
        return expect(isValidFunction(value));
    };

    const changeInput = (label: string, value: string) => {
        const input = wrapper.find(`Input[label='${label}']`).first();
        const onChangeFn = input.prop<(v: string) => void>('onChange');
        onChangeFn(value);
    };
});