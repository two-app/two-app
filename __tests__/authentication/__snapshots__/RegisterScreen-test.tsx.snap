// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should maintain snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": "10%",
            "marginRight": "10%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#EE7674",
              "fontSize": 50,
              "fontWeight": "bold",
              "paddingBottom": 20,
              "paddingTop": 20,
              "textAlign": "center",
            }
          }
        >
          two.
        </Text>
        <Text
          data-testid="heading"
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 25,
              "fontWeight": "bold",
            }
          }
        >
          Sign Up
        </Text>
        <Text
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          First Name
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCompleteType="name"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="First Name"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderBottomColor": "rgba(0, 0, 0, 0.2)",
                "borderBottomWidth": 1,
                "fontSize": 15,
                "paddingBottom": 10,
                "paddingHorizontal": 0,
                "paddingTop": 10,
                "width": "100%",
              },
              undefined,
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <Text
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Last Name
        </Text>
        <TextInput
          allowFontScaling={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Last Name"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderBottomColor": "rgba(0, 0, 0, 0.2)",
                "borderBottomWidth": 1,
                "fontSize": 15,
                "paddingBottom": 10,
                "paddingHorizontal": 0,
                "paddingTop": 10,
                "width": "100%",
              },
              undefined,
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <Text
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Email
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCompleteType="email"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="you@email.com"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "borderBottomColor": "rgba(0, 0, 0, 0.2)",
                "borderBottomWidth": 1,
                "fontSize": 15,
                "paddingBottom": 10,
                "paddingHorizontal": 0,
                "paddingTop": 10,
                "width": "100%",
              },
              undefined,
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <Text
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Password
        </Text>
        <TextInput
          allowFontScaling={true}
          autoCompleteType="password"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Secure Password"
          rejectResponderTermination={true}
          secureTextEntry={true}
          style={
            Array [
              Object {
                "borderBottomColor": "rgba(0, 0, 0, 0.2)",
                "borderBottomWidth": 1,
                "fontSize": 15,
                "paddingBottom": 10,
                "paddingHorizontal": 0,
                "paddingTop": 10,
                "width": "100%",
              },
              undefined,
              undefined,
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(0, 0, 0, 0.2)",
              "justifyContent": "center",
              "marginBottom": 25,
              "marginTop": 25,
              "opacity": 1,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontWeight": "bold",
              }
            }
          >
            Join Two
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
